### ISSUE TITLE
Probing of Latent Space Learned By SAM

### ISSUE THREAD

**Issue by aymane399** [facebookresearch/segment-anything#283]
Hello and thank you for sharing your work!

Could you please specify more precisely how tocompute the "mask embeddings" refered in  "D.6. Probing the Latent Space of SAM".
In particular the multiplication between a binary mask and the image embedding of the crop around the mask, and the signification of spatial averaging of resulting product.

Thanks

**Comment by lizhogn** [comment]
I have a similar confusion and I am trying to perform a one-shot task. As shown in the figure below, if I am given a mask of a tire, I can extract the corresponding Embedding of this mask and take the average in the spatial dimension to obtain a (1, 256, 1, 1) shaped Embedding. I then attempt to match this Embedding with the full-image features (shape [1, 256, 64, 64]) using similarity matching. However, the result, as shown in the similarity, indicates that I cannot effectively extract the position of another tire. Can you please explain how can it works in "D.6. Probing the Latent Space of SAM"?

![image](https://user-images.githubusercontent.com/30550546/233992902-0336cf81-b512-46a5-ad89-28c5a5aa4eec.png)

Interestingly, if I do not perform spatial averaging on the Embedding corresponding to the mask and directly convolve it with the full-image features, it seems to be feasible (though it not aways works):

![image](https://user-images.githubusercontent.com/30550546/233995325-696a1135-785c-494d-9997-899b98bf17d1.png)

**Comment by Animadversio** [comment]
My way of doing it is a bit hacky, not sure if it's the way Meta people do it. 
Basically averaging the embedding within the segmentation mask, after resizing. 

```python 
def get_mask_embed(mask, img_embed):
    """Extract mask embedding

    :param mask: H x W, binary mask in numpy
    :param img_embed: 1 x c x h x w, image embedding tensor [256, 64, 64]
    :return:
        query_embed: 1 x c, mask embedding tensor
        mask_resize: 1 x 1 x h x w, the mask that is used to extract the embedding
    """
    orig_H, orig_W = mask.shape[:2]
    embed_H, embed_W = img_embed.shape[-2:]
    if orig_H >= orig_W:
        resize_W = int(embed_H * orig_W / orig_H)
        resize_H = embed_H
    else:
        resize_H = int(embed_W * orig_H / orig_W)
        resize_W = embed_W
    mask_resize = interpolate(torch.tensor(mask)[None, None].float(),
                size=(resize_H, resize_W), mode='nearest')
    query_embed = (img_embed[:, :, :resize_H, :resize_W] * mask_resize.cuda()).sum(dim=(-2, -1)) / mask_resize.sum()
    return query_embed, mask_resize
```

Some demo of this is like the following. 


My question is using this kind of method, the mask embedding will contain the location information of the mask (position embedding info), not sure if it's helpful or not counter-helpful. Because I want to find similar masks all over the place instead of just around the query mask.

**Comment by SpencerWhitehead** [comment]
Hello, thanks for your interest! 

We are not intending to release the latent space probing as part of this repo at this time. In general, for research-related issues, please reach out to the authors directly instead of using github issues.

Here are some details that may be relevant:
- Image embeddings here come directly from the ViT backbone, before the channel reducing conv layers.
- We mention in the paper that we crop the image centered around the mask to compute mask embeddings. For this, we take the bounding box for the mask, expand its size by a factor (e.g., 2.0 or 1.5), and use the bounding box region as our cropped image.
- Given the image embedding for the cropped image, we resize it back to the input resolution to multiply by the mask for the crop. We average over the remaining non-zero elements.
- We also do this `cropping -> embedding * mask -> averaging` process on a h-flipped version of the image and then average the original and h-flipped mask embeddings to get the final mask embedding.
- We use cosine similarity as our similarity metric.
